napari/layers/shapes/_shapes_mouse_bindings.py

```diff
@@ -0,0 +1,257 @@
+import numpy as np
+from copy import copy
+from ._shapes_constants import Mode
+from ._shapes_models import Rectangle, Ellipse, Line, Path, Polygon
+from ._shapes_utils import point_to_lines
+
+
+def highlight(layer, event):
+    """Highlight hovered shapes."""
+    layer._set_highlight()
+
+
+def select(layer, event):
+    """Select shapes or vertices either in select or direct select mode.
+
+    Once selected shapes can be moved or resized, and vertices can be moved
+    depending on the mode. Holding shift when resizing a shape will preserve
+    the aspect ratio.
+    """
+    shift = 'Shift' in event.modifiers
+    # on press
+    layer._moving_value = copy(layer._value)
+    if layer._value[1] is None:
+        if shift and layer._value[0] is not None:
+            if layer._value[0] in layer.selected_data:
+                layer.selected_data.remove(layer._value[0])
+                shapes = layer.selected_data
+                layer._selected_box = layer.interaction_box(shapes)
+            else:
+                layer.selected_data.append(layer._value[0])
+                shapes = layer.selected_data
+                layer._selected_box = layer.interaction_box(shapes)
+        elif layer._value[0] is not None:
+            if layer._value[0] not in layer.selected_data:
+                layer.selected_data = [layer._value[0]]
+        else:
+            layer.selected_data = []
+    layer._set_highlight()
+    yield
+
+    # on move
+    while event.type == 'mouse_move':
+        # Drag any selected shapes
+        coord = [layer.coordinates[i] for i in layer.dims.displayed]
+        layer._move(coord)
+        yield
+
+    # on release
+    shift = 'Shift' in event.modifiers
+    if not layer._is_moving and not layer._is_selecting and not shift:
+        if layer._value[0] is not None:
+            layer.selected_data = [layer._value[0]]
+        else:
+            layer.selected_data = []
+    elif layer._is_selecting:
+        layer.selected_data = layer._data_view.shapes_in_box(layer._drag_box)
+        layer._is_selecting = False
+        layer._set_highlight()
+    layer._is_moving = False
+    layer._drag_start = None
+    layer._drag_box = None
+    layer._fixed_vertex = None
+    layer._moving_value = (None, None)
```

Comment:
where possible, could you convert these to local variables instead of putting them on the layer?

Reply:
Unfortunately all of these occur in the main shapes model too right now. The biggest offender is the `layer._move` function. That could probably be pulled in here to be local too, but that would be a much bigger lift and I'd like to save that for a follow up PR